VMware: new module vmware_guest_serial_port

anusha94 opened the pull request:
##### SUMMARY
- Fixes #54956 
- The ask was to add an option in the vmware_guest itself
- Upon inspection, a serial port can be created in multiple ways and so can be a module in itself
- Therefore, created a new module called vmware_serial_port


##### ISSUE TYPE

- New Module Pull Request

##### COMPONENT NAME
vmware_serial_port

##### ADDITIONAL INFORMATION
This module supports serial port with 4 different backing types

- network backing
- pipe backing
- file backing
- physical serial port backing

Testing done:

- sanity
- integration with govcsim
- pycodestyle

anusha94 committed:
Adding a new module to create serial port and add it to VM
Signed-off-by: Anusha Hegde

anusha94 committed:
Adding assertions to main.yml
Signed-off-by: Anusha Hegde

anusha94 commented:
@devdil @pgbidkar
Please review and provide your feedback.

Thanks!

Akasurde commented on lib/ansible/modules/cloud/vmware/vmware_serial_port.py:
```diff
@@ -0,0 +1,345 @@
+
+DOCUMENTATION = '''
+---
+module: vmware_serial_port
```
I would suggest to rename this module to `vmware_guest_serial_port` in order to match other modules in VMware space.

Akasurde commented on lib/ansible/modules/cloud/vmware/vmware_serial_port.py:
```diff
@@ -0,0 +1,345 @@
+---
+module: vmware_serial_port
+
+short_description: Create a serial port
```
```suggestion
short_description: Manage serial ports on an existing VM
```

Akasurde commented on lib/ansible/modules/cloud/vmware/vmware_serial_port.py:
```diff
@@ -0,0 +1,345 @@
+
+short_description: Create a serial port
+
+version_added: "1.0"
```
```suggestion
version_added: "2.10"
```

Akasurde commented on lib/ansible/modules/cloud/vmware/vmware_serial_port.py:
```diff
@@ -0,0 +1,345 @@
+version_added: "1.0"
+
+description:
+    - "Creates a serial port on an existing VM"
```
```suggestion
    - This module can be used to manage serial ports on an existing virtual machine.
```

Akasurde commented on lib/ansible/modules/cloud/vmware/vmware_serial_port.py:
```diff
@@ -0,0 +1,345 @@
+    - "Creates a serial port on an existing VM"
+
+options:
+    hostname:
```
Please use `extends_documentation_fragment: vmware.documentation` instead of specifying `hostname`, `username` and `password`

Akasurde commented on lib/ansible/modules/cloud/vmware/vmware_serial_port.py:
```diff
@@ -0,0 +1,345 @@
+        required: true
+    name:
+        description:
+            - The name of the Virtual Machine
```
```suggestion
            - The name of the Virtual Machine.
```

Add full stop at end of each statement in description.

Akasurde commented on lib/ansible/modules/cloud/vmware/vmware_serial_port.py:
```diff
@@ -0,0 +1,345 @@
+        type: str
+    vm_id:
+        description:
+            - The id of the Virtual Machine
```
Could you please add description on which ID to specify - `instance_uuid` or `BIOS_uuid` ? also add [`use_instance_uuid`](https://github.com/ansible/ansible/blob/a99af710fbfb768745b214f1a4de2f0dd5d40551/lib/ansible/modules/cloud/vmware/vmware_guest_boot_manager.py#L49) parameter to match other modules.

Akasurde commented on lib/ansible/modules/cloud/vmware/vmware_serial_port.py:
```diff
@@ -0,0 +1,345 @@
+'''
+
+RETURN = r'''
+#
```
Please add return output (This gives user a better idea about expected output and format).

Akasurde commented on lib/ansible/modules/cloud/vmware/vmware_serial_port.py:
```diff
@@ -0,0 +1,345 @@
+import time
+from ansible.module_utils.basic import AnsibleModule
+from ansible.module_utils.vmware import PyVmomi, vmware_argument_spec, set_vm_power_state, wait_for_task, gather_vm_facts
+from pyVmomi import vim
```
```suggestion
try:
    from pyVmomi import vim
except ImportError:
    pass
```

Akasurde commented on lib/ansible/modules/cloud/vmware/vmware_serial_port.py:
```diff
@@ -0,0 +1,345 @@
+          - vm: Virtual Machine
+
+        Output:
+          - [proceed, current_state]
```
Use tuple instead of list

Akasurde commented on lib/ansible/modules/cloud/vmware/vmware_serial_port.py:
```diff
@@ -0,0 +1,345 @@
+        else:
+            return {'changed': self.change_applied, 'failed': False, 'msg': "Serial port successfully created"}
+
+    def wait_for_task(self, task, poll_interval=1):
```
Please use [`wait_for_task`](https://github.com/ansible/ansible/blob/a99af710fbfb768745b214f1a4de2f0dd5d40551/lib/ansible/module_utils/vmware.py#L53)

Akasurde commented on lib/ansible/modules/cloud/vmware/vmware_serial_port.py:
```diff
@@ -0,0 +1,345 @@
+
+    pyv = PyVmomiHelper(module)
+    # Check if the VM exists before continuing
+    print("checking vm")
```
Remove debugging print statement.

Akasurde commented on lib/ansible/modules/cloud/vmware/vmware_serial_port.py:
```diff
@@ -0,0 +1,345 @@
+        else:
+            module.fail_json(msg="The attempted operation cannot be performed in the current state (" + current_state + "), use the force option to forcefully power off the VM")
+            '''
+            result.update(
```
Remove commented code.

Akasurde commented on lib/ansible/modules/cloud/vmware/vmware_serial_port.py:
```diff
@@ -0,0 +1,345 @@
+
+
+def main():
+    argument_spec = vmware_argument_spec()
```
Please add capabilities like `remove` and `modify` of serial port.

Akasurde commented on lib/ansible/modules/cloud/vmware/vmware_serial_port.py:
```diff
@@ -0,0 +1,345 @@
+    argument_spec.update(
+        uuid=dict(type='str'),
+        vm_id=dict(type='str'),
+        backing=dict(type='str', choices=['network', 'file', 'pipe', 'device']),
```
Could you please consider handling multiple serial port ? Please take a look at `vmware_guest_network` module for such functionality where it handles multiple network devices at single time.

Akasurde reviewed (changes requested):
Thanks for the contribution.

anusha94 committed:
Resolved review comments
Added support for multiple serial ports
Added remove serial port functionality
Added modify serial port functionality

Signed-off-by: Anusha Hegde

anusha94 commented:
@Akasurde 
Thanks for the detailed review. I have made the changes, please review again.

Akasurde commented on lib/ansible/modules/cloud/vmware/vmware_guest_serial_port.py:
```diff
@@ -0,0 +1,458 @@
+    type: str
+  uuid:
+    description:
+      - UUID of the instance to gather info if known, this is VMware's unique identifier.
```
```suggestion
      - UUID of the instance to manage the serial ports, this is VMware's unique identifier.
```

Akasurde commented on lib/ansible/modules/cloud/vmware/vmware_guest_serial_port.py:
```diff
@@ -0,0 +1,458 @@
+    returned: always
+    type: dict
+    sample: {
+        "0": {
```
I find a list of dict easier to handle rather than dict of dict.

Akasurde commented on lib/ansible/modules/cloud/vmware/vmware_guest_serial_port.py:
```diff
@@ -0,0 +1,458 @@
+    """
+    Get the serial port info
+    """
+    serial_port_info = dict()
```
```suggestion
    serial_port_info = list()
```

Akasurde commented on lib/ansible/modules/cloud/vmware/vmware_guest_serial_port.py:
```diff
@@ -0,0 +1,458 @@
+                backing['file_path'] = port.backing.fileName
+            else:
+                continue
+            serial_port_info[port_index] = backing
```
```suggestion
            backing['port_index'] = port_index
            serial_port_info.append(backing)
```

Akasurde commented on lib/ansible/modules/cloud/vmware/vmware_guest_serial_port.py:
```diff
@@ -0,0 +1,458 @@
+    """
+    serial_spec = vim.vm.device.VirtualDeviceSpec()
+    serial_port = vim.vm.device.VirtualSerialPort()
+    serial_port.yieldOnPoll = True
```
I maybe wrong but we are getting user input for `yield_on_poll` but you are not using here.

Akasurde commented on lib/ansible/modules/cloud/vmware/vmware_guest_serial_port.py:
```diff
@@ -0,0 +1,458 @@
+    username: '{{ vcenter_username }}'
+    password: '{{ vcenter_password }}'
+    name: '{{ name }}'
+    backing: network
```
Need to update example with respect to new list parameter.

Akasurde commented on test/integration/targets/vmware_guest_serial_port/tasks/main.yml:
```diff
@@ -0,0 +1,113 @@
+      - type: 'file'
+        state: 'new'
+        file_path: 'newfile'
+      yield_on_poll:  True
```
```suggestion
        yield_on_poll:  True
```

Akasurde reviewed (changes requested):
Thanks for the update.

anusha94 committed:
Resolved review comments
Signed-off-by: Anusha Hegde

Akasurde reviewed (approved):
LGTM, shipit

goneri commented on test/integration/targets/vmware_guest_serial_port/tasks/main.yml:
```diff
@@ -0,0 +1,113 @@
+    vars:
+      setup_attach_host: true
+      setup_datastore: true
+      setup_virtualmachines: true
```
Since you manually create a VM after, you don't need to create one with `prepare_vmware_tests`. Please, remove the `setup_virtualmachines: true`.

goneri commented on test/integration/targets/vmware_guest_serial_port/tasks/main.yml:
```diff
@@ -0,0 +1,113 @@
+# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
+
+
+- when: vcsim is defined
```
Some functional tests cannot be run on govcsim (a vcenter API simulator), in this case, we start a block with `when: vcsim is not defined`.
There is no reason to do the opposite, we should be able to run the tests on a regular lab without any problem. Could you please remove the condition and the associated block.

goneri commented on test/integration/targets/vmware_guest_serial_port/tasks/main.yml:
```diff
@@ -0,0 +1,113 @@
+      name: test_vm1
+      guest_id: centos64Guest
+      datacenter: "{{ dc1 }}"
+      folder: F0
```
Please use '{{ f0 }}' instead of F0.

goneri commented on lib/ansible/modules/cloud/vmware/vmware_guest_serial_port.py:
```diff
@@ -0,0 +1,439 @@
+        facts = gather_vm_facts(self.content, vm_obj)
+        current_state = facts['hw_power_status'].lower()
+        if current_state == "poweredoff":
+            return [True, current_state]
```
I would prefer you just raise an error here and ask the user to shut down the VM with `vmware_guest_powerstate`.

Each Ansible should handle just one operation, this simplify the documentation and the maintenance.

goneri commented on lib/ansible/modules/cloud/vmware/vmware_guest_serial_port.py:
```diff
@@ -0,0 +1,439 @@
+        """
+        # create serial config spec for adding, editing, removing
+        for backing in self.params.get('backings'):
+            if backing['state'].lower() == 'new':
```
I don't like the `new` state because it breaks the idempotency of the module. It sounds like you can use `service_uri`, `device_name` and `file_path` to know if a device is already present. Could you do that instead?

anusha94 replied:
@goneri 
How about the usecase when the user wants to attach 2 serial devices of same backing type?

goneri replied:
I believe it should be type+device_name/service_uri/file_path.

goneri commented on lib/ansible/modules/cloud/vmware/vmware_guest_serial_port.py:
```diff
@@ -0,0 +1,439 @@
+        uuid=dict(type='str'),
+        moid=dict(type='str'),
+        use_instance_uuid=dict(type='bool', default=False),
+        force=dict(type='bool', default=False),
```
As explain above, could you please remove this argument?

goneri commented on test/integration/targets/vmware_guest_serial_port/tasks/main.yml:
```diff
@@ -0,0 +1,113 @@
+      datacenter: "{{ dc1 }}"
+      folder: F0
+      hardware:
+        num_cpus: 4
```
In order to reduce the CI requirement, could you just use 1 vcpu?

goneri reviewed (changes requested):
Hi @anusha94,

Thank you for the module and welcome in the Ansible community :-) There is a couple of little things that I would prefer to see fixed before the inclusion. They will probably be sources of problem in the future, and it will be harder to improve the module once it will be released. This is the reason why I would prefer to tackle them now.

By the way, this is a nice first module, congratulation :-)

anusha94 committed:
Removed the new option for creating a serial port, instead check if already present, else create a new one
Removed the govcsim block
Addressed other review comments

Signed-off-by:  Anusha Hegde

anusha94 committed:
fixed invalid yml error

Signed-off-by: Anusha Hegde

anusha94 commented:
@goneri ,
Thank you for the review :) Please verify the changes.

goneri commented on test/integration/targets/vmware_guest_serial_port/tasks/main.yml:
```diff
@@ -0,0 +1,100 @@
+# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
+
+- import_role:
+    name: prepare_vmware_tests
```
You need to add:

```
  vars:
    setup_attach_host: true
    setup_datastore: true
```

- setup_attach_host: the ESXi will be attached to the vCenter
- setup_datastore: datastore are mounted on the ESXi

goneri commented:
Thanks @anusha94. It's much better without the `state: new`. I actually would like the same change in `vmware_guest_network` :-): https://github.com/ansible/ansible/issues/63648

goneri commented on test/integration/targets/vmware_guest_serial_port/tasks/main.yml:
```diff
@@ -0,0 +1,100 @@
+    that:
+        - create_vm_for_test is changed
+
+- name: Create multiple serial ports with Backing type - network, pipe, device and file
```
This step fails with the following error:

```
TASK [vmware_guest_serial_port : Create multiple serial ports with Backing type - network, pipe, device and file] ****************************************************************************
task path: /home/goneri/.ansible/test/tmp/vmware_guest_serial_port-_6233y53-ÅÑŚÌβŁÈ/test/integration/targets/vmware_guest_serial_port/tasks/main.yml:36
<testhost> ESTABLISH LOCAL CONNECTION FOR USER: goneri
<testhost> EXEC /bin/sh -c 'echo ~goneri && sleep 0'
<testhost> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo /home/goneri/.ansible/tmp/ansible-tmp-1571348205.6692333-209565769524555 `" && echo ansible-tmp-1571348205.6692333-209565769524555="` echo /home/goneri/.ansible/tmp/ansible-tmp-1571348205.6692333-209565769524555 `" ) && sleep 0'                                                                                             
Using module file /home/goneri/git_repos/ansible_projects/b/lib/ansible/modules/cloud/vmware/vmware_guest_serial_port.py
<testhost> PUT /home/goneri/.ansible/tmp/ansible-local-27978cc47up40/tmpy1djlbkh TO /home/goneri/.ansible/tmp/ansible-tmp-1571348205.6692333-209565769524555/AnsiballZ_vmware_guest_serial_port.py                                                                                                                                                                                          
<testhost> EXEC /bin/sh -c 'chmod u+x /home/goneri/.ansible/tmp/ansible-tmp-1571348205.6692333-209565769524555/ /home/goneri/.ansible/tmp/ansible-tmp-1571348205.6692333-209565769524555/AnsiballZ_vmware_guest_serial_port.py && sleep 0'                                                                                                                                                  
<testhost> EXEC /bin/sh -c 'ESXI1_HOSTNAME=esxi3.test ESXI1_PASSWORD=root ESXI1_USERNAME=root ESXI2_HOSTNAME=test4.test ESXI2_PASSWORD=root ESXI2_USERNAME=root VCENTER_HOSTNAME=vcenter.test VCENTER_PASSWORD='"'"'!234AaAa56'"'"' VCENTER_USERNAME=administrator@vsphere.local VMWARE_VALIDATE_CERTS=false /tmp/python-wq6lw5ae-ansible/python /home/goneri/.ansible/tmp/ansible-tmp-1571348205.6692333-209565769524555/AnsiballZ_vmware_guest_serial_port.py && sleep 0'                                                                                                               
<testhost> EXEC /bin/sh -c 'rm -f -r /home/goneri/.ansible/tmp/ansible-tmp-1571348205.6692333-209565769524555/ > /dev/null 2>&1 && sleep 0'
fatal: [testhost]: FAILED! => {
    "changed": false,
    "module_stderr": "pyVmomi.VmomiSupport.InvalidDatastorePath: (vim.fault.InvalidDatastorePath) {\n   dynamicType = <unset>,\n   dynamicProperty = (vmodl.DynamicProperty) [],\n   msg = \"Invalid datastore path 'newfile'.\",\n   faultCause = <unset>,\n   faultMessage = (vmodl.LocalizableMessage) [],\n   datastore = <unset>,\n   name = <unset>,\n   datastorePath = 'newfile'\n}\n\nThe above exception was the direct cause of the following exception:\n\nTraceback (most recent call last):\n  File \"/home/goneri/.ansible/tmp/ansible-tmp-1571348205.6692333-209565769524555/AnsiballZ_vmware_guest_serial_port.py\", line 116, in <module>\n    _ansiballz_main()\n  File \"/home/goneri/.ansible/tmp/ansible-tmp-1571348205.6692333-209565769524555/AnsiballZ_vmware_guest_serial_port.py\", line 108, in _ansiballz_main\n    invoke_module(zipped_mod, temp_path, ANSIBALLZ_PARAMS)\n  File \"/home/goneri/.ansible/tmp/ansible-tmp-1571348205.6692333-209565769524555/AnsiballZ_vmware_guest_serial_port.py\", line 54, in invoke_module\n    runpy.run_module(mod_name='ansible.modules.cloud.vmware.vmware_guest_serial_port', init_globals=None, run_name='__main__', alter_sys=False)\n  File \"/usr/lib64/python3.7/runpy.py\", line 208, in run_module\n    return _run_code(code, {}, init_globals, run_name, mod_spec)\n  File \"/usr/lib64/python3.7/runpy.py\", line 85, in _run_code\n    exec(code, run_globals)\n  File \"/tmp/ansible_vmware_guest_serial_port_payload_wgxarmo1/ansible_vmware_guest_serial_port_payload.zip/ansible/modules/cloud/vmware/vmware_guest_serial_port.py\", line 454, in <module>\n  File \"/tmp/ansible_vmware_guest_serial_port_payload_wgxarmo1/ansible_vmware_guest_serial_port_payload.zip/ansible/modules/cloud/vmware/vmware_guest_serial_port.py\", line 438, in main\n  File \"/tmp/ansible_vmware_guest_serial_port_payload_wgxarmo1/ansible_vmware_guest_serial_port_payload.zip/ansible/modules/cloud/vmware/vmware_guest_serial_port.py\", line 242, in reconfigure_vm_serial_port\n  File \"/tmp/ansible_vmware_guest_serial_port_payload_wgxarmo1/ansible_vmware_guest_serial_port_payload.zip/ansible/module_utils/vmware.py\", line 82, in wait_for_task\n  File \"<string>\", line 3, in raise_from\nansible.module_utils.vmware.TaskError: (\"Invalid datastore path 'newfile'.\", None)\n",                                                                                                                                                                                       
    "module_stdout": "",
    "msg": "MODULE FAILURE\nSee stdout/stderr for the exact error",
    "rc": 1
}
```

anusha94 replied:
Sorry for the delayed response.

The error was due to filename, it should be of format `<datastore>/<filename>`

I have updated the same in the documentation, examples and integration test.

anusha94 committed:
Updated example for creating serial port of backing type file

Signed-off-by: Anusha Hegde

anusha94 committed:
added vars in integration test

Signed-off-by: Anusha Hegde

anusha94 committed:
changed datastore name in yml

Signed-off-by: Anusha Hegde

anusha94 committed:
Changed the logic to add multiple serial ports
There was an error due to the same device type for all the serial ports

Updated integration test and documentation

pgbidkar commented:
Hi @goneri @Tomorrow9 , Please can you review this module?

Thanks

goneri commented on test/integration/targets/vmware_guest_serial_port/tasks/main.yml:
```diff
@@ -0,0 +1,103 @@
+    - type: 'device'
+      device_name: '/dev/char/serial/uart0'
+    - type: 'file'
+      file_path: '[datastore1]/file1'
```
`file_path: '[ {{ ds2 }} ]/file1'` or if https://github.com/ansible/ansible/pull/64783 is merged: `file_path: '[ {{ rw_datastore}} ]/file1'`

anusha94 replied:
used `file_path: '[ {{ rw_datastore}} ]/file1'` hoping #64783 will be merged soon :)

anusha94 replied:
had to use `ds2` only, else integration test was failing, probably in future once your PR is merged I can change it

goneri reviewed (changes requested):
Just one last little problem and the PR can be merge. Thank you for your work :-)

anusha94 committed:
changed datastore name in main.yml

Signed-off-by: Anusha Hegde

Tomorrow9 commented on lib/ansible/modules/cloud/vmware/vmware_guest_serial_port.py:
```diff
@@ -0,0 +1,469 @@
+        self.config_spec.deviceChange = []
+        self.serial_ports = []
+
+    def check_vm_state(self, vm_obj):
```
Hi, is this function can be simplified to 'if vm_obj.runtime.powerState != vim.VirtualMachinePowerState.poweredOff' as what used in other modules? Thanks.

anusha94 replied:
removed `gather_vm_facts` function call

Thanks for pointing out :)

Tomorrow9 commented on lib/ansible/modules/cloud/vmware/vmware_guest_serial_port.py:
```diff
@@ -0,0 +1,469 @@
+        return serial_spec
+
+
+def get_serial_port(vm_obj, backing):
```
Hi, is there any reason that put these two functions outside the class? Thanks.

anusha94 replied:
@Tomorrow9 because there are no references to class members in this function, thought it can be outside the class.

anusha94 committed:
changed datastore value
removed gather_vm_facts function call

Signed-off-by: Anusha Hegde

goneri commented:
Thanks @anusha94
